packages/web-components/fast-foundation/src/tree-item/tree-item.ts

```diff
@@ -46,6 +51,11 @@ export class TreeItem extends FoundationElement {
      */
     @attr({ mode: "boolean" })
     public selected: boolean;
+    private selectedChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.$emit("selected-change", this);
```

Comment:
I *think* this behavior is inconsistent with the existing behavior, in that this event now emits for `true` -> `false` as well as `false` -> `true` mutations. I believe the existing implementation only emits for `false` -> `true` mutations (I could be mistaken here, and if so please tell me :)). Is the deviation here intentional?

Reply:
Not quite.  Since a recent change the tree item emits "selected-change" on click, every click, before the parent tree actually changes the 'selected' prop of the item.  If a dev hooked up a listener on that item they could easily get the change event prior to the state actually changing which is not ideal.  That being said, should we only be emitting for true?  I don't see an upside to that myself.

Reply:
Yea I understand this is mostly to fix timing, but my point is this also introduces a new behavior - that is the part I'm wary of. I don't know if we should only be emitting for true, but emitting for false will be behaviorally different. IMO fixing a regression to the event timing isn't "breaking". Adding event emission in the opposite selection case seems to me to be "breaking".

Reply:
I don't think that's the case.  This is the handler invoked on click/keydown from months ago, I didn't filter out deselection.

`` private handleSelected(e?: Event): void {
        this.selected = !this.selected;
        this.$emit("selected-change", e);
    }``

Reply:
This is the code prior to the timing regression: https://github.com/microsoft/fast/blob/3ad50c8eec98b3156a4559586ca76e4abebefad8/packages/web-components/fast-foundation/src/tree-item/tree-item.ts#L299

That handler is only invoked on item click and "enter" key input though, not whenever `selected` is changed as this code is.

Reply:
The initial implementation changed the selected prop and emitted a "selected-change" event from the tree item on click/keystroke.  The parent tree listened for and reacted to that event.  The change last month modified that such that the decision as to whether a clicked element should become selected or not was made at the tree level as a result of the same "selected-change" event as previous, but now the event was emitted prior to the actual state change.  The proposal here is to have the tree listen for click events directly and set selection based on that, which in turn emits the "selected-change" event based on the property change.  This also means that the tree can handle developers setting the selected state of an item directly.

Reply:
I think it's reasonable to argue that the original implementation emitted the event on property change just as much as on user interaction as the two were closely meshed.

Reply:
I understand the goals. Timing aside, my pushback is specific to the fact that this pull-request causes `tree-item`s to emit a `selected-change` event both when becoming *selected* and *deselected*, whereas the existing behavior is that they only emit a `selected-change` event when becoming *selected*.

Reply:
Actually I guess that's not quite true. It would emit for deselection in cases of user-interaction of the item itself.

Reply:
> I understand the goals. Timing aside, my pushback is specific to the fact that this pull-request causes `tree-item`s to emit a `selected-change` event both when becoming _selected_ and _deselected_, whereas the existing behavior is that they only emit a `selected-change` event when becoming _selected_.

Ok, so there is a subtle difference. It is not that a tree item that is clicked on while selected now emits a "selection-changed" event for becoming deselected, but rather that items who become deselected for reasons other than being clicked on directly would now emit as well.

Reply:
Hmm yea that makes sense. It's still not my favorite because everyone listening for these events on some parent like the tree-view is going to have to do work to determine which item is being selected and which is being deselected, and that the handler will be invoked twice for each *act of selection*, but theoretically they should already be doing that. I no longer think it's breaking though.

Reply:
I'm thinking we should add a selection changed event on the tree view for folks who don't want don't want to be so atomic about it.  Ie.  we'd emit it once when the selected item has changed and all the child item state is set.

Reply:
@scomea, just so we're clean, it would now be considered an anti-pattern (or, more aptly, _incorrect_) to, in a derived `TreeItem`, handle the `selected-change` event and change the current `selected` state to something else. The reason for doing this would be to prevent a user from de-selecting a currently selected tree item (by clicking on it).

Instead, it seems like the correct implementation for this behavior would be to override the `click` event, update the `selected` state as necessary, and stop the propagation of it if needed. Is this right?

Reply:
@atmgrifter00 I generally consider calling stopPropagation to bad form and would call preventDefault on the initial click event instead.  The parent tree should ignore click events marked as default prevented. But yes, authors should be able to listen from click events on tree items, handle it with their own code and set the selected state of tree items as they see fit. The parent tree should update to reflect the selected state changes on child tree items.